site/js/modules/webui/input.js

```diff
@@ -0,0 +1,187 @@
+
+        // The placeholder text for the input.
+        this.placeholder = placeholder;
+
+        // Optional classes that are attached to the input.
+        this.inputClass = inputClass;
+
+        // Any additional attributes to the input field.
+        // If the field is required, the required attribute will be added automatically.
+        this.attributes = attributes;
+
+        // Determines the position of the HTML label with respect to the input.
+        this.labelBefore = labelBefore;
+
+        // A custom function for extracting the value from an input.
+        // If a default value extraction cannot be inferred, the value is parsed using JSON.
+        this.extractInputFunc = extractInputFunc;
+    }
+
+    validate() {
+        if ((this.name == undefined) || (this.name == '')) {
+            console.error(`Input field cannot have an empty name: ${JSON.stringify(this)}.`);
+        }
+
+        if ((this.displayName == undefined) || (this.displayName == '')) {
+            console.error(`Input field cannot have an empty display name: ${JSON.stringify(this)}.`);
+        }
+    }
+
+    toHTML() {
+        this.validate();
+
+        let attributes = this.attributes;
+        let displayName = this.displayName;
+        if ((this.required) && (this.placeholder === "")) {
+            attributes += ' required';
+            displayName += ` <span class="required-color">*</span>`;
+        }
+
+        let inputFieldHTML = [
+            `<label for="${this.name}">${displayName}</label>`,
+            `<input type="${this.type}" id="${this.name}" name="${this.name}" placeholder="${this.placeholder}" ${attributes}/>`,
+        ];
+
+        if (!this.labelBefore) {
+            inputFieldHTML.reverse();
+        }
+
+        return `
+            <div class="input-field ${this.inputClass}">
+                ${inputFieldHTML.join("\n")}
+            </div>
+        `;
+    }
+
+    getResult(container) {
+        let input = container.querySelector(`fieldset [name=${this.name}]`);
+        input.classList.add("touched");
+
+        return new Result(input, this.underlyingType, this.extractInputFunc);
+    }
+}
+
+// The result of getting an Input.Field.
+class Result {
+    constructor(input, underlyingType, extractInputFunc = undefined) {
+        // The input from the query selector.
+        this.input = input;
+
+        // See Field for field descriptions.
+        this.underlyingType = underlyingType;
+        this.extractInputFunc = extractInputFunc;
+    }
+
+    // Validate the value of the input.
+    // Throws an error on invalid input values.
+    validate() {
+        if (!this.input.validity.valid) {
+            throw new Error(`${this.input.validationMessage}`);
+        }
+
+        // Skip further validation if a custom extraction function is provided.
+        if (this.extractInputFunc != undefined) {
+            return;
+        }
+
+        if (!this.input || this.input.value === "") {
```

Comment:
Is `this.input` allowed to be undefined?
Should this be checked in the constructor?